pkgs/by-name/ld/ld64/0002-Avoid-UB-when-std-vector-is-empty.patch

```diff
@@ -0,0 +1,1239 @@
+From e4434c35b2922bbcf1034ccfc985c33cf6afc90b Mon Sep 17 00:00:00 2001
```

Comment:
if this is not preventing compile failures or known bugs it should not be part of the change, it just seems to add to maintenance burden -- fixing all code to align with the c++ std isn't a requirement for pkgs, AFAIK

Reply:
Without the patch, ld64 crashes when trying to link itself later in the bootstrap.

Reply:
on both aarch64 and x64?

Reply:
aarch64-darwin. This crash happened every time. I didn’t check x86_64-darwin. Those crashes happen with the patch non-deterministically (i.e., you can retry the build, and it succeeds).

```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               ld [8606]
Path:                  /private/tmp/*/ld
Identifier:            ld
Version:               ???
Code Type:             ARM-64 (Native)
Parent Process:        clang-18 [8600]
Responsible:           nix [24253]
User ID:               301

Date/Time:             2024-07-01 02:25:42.8480 -0400
OS Version:            macOS 14.5 (23F79)
Report Version:        12
Anonymous UUID:        8A4D7F20-D75B-4FA7-A8E5-8288BDD004AC

Sleep/Wake UUID:       78BE2690-BB1C-4CF7-9F3B-06B2990D7233

Time Awake Since Boot: 1100000 seconds
Time Since Wake:       22244 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BREAKPOINT (SIGTRAP)
Exception Codes:       0x0000000000000001, 0x000000010071d37c

Termination Reason:    Namespace SIGNAL, Code 5 Trace/BPT trap: 5
Terminating Process:   exc handler [8606]

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   ld                            	       0x10071d37c mach_o::relocatable::Atom<arm64>::fixupsEnd() const + 60
1   ld                            	       0x1006aa9b0 ld::tool::Resolver::convertReferencesToIndirect(ld::Atom const&) + 60
2   ld                            	       0x1006aa5ac ld::tool::Resolver::doAtom(ld::Atom const&) + 1224
3   ld                            	       0x10071a378 mach_o::relocatable::File<arm64>::forEachAtom(ld::File::AtomHandler&) const + 84
4   ld                            	       0x1005c3c20 ld::tool::InputFiles::forEachInitialAtom(ld::File::AtomHandler&, ld::Internal&) + 620
5   ld                            	       0x1006b0468 ld::tool::Resolver::resolve() + 152
6   ld                            	       0x1006c73ec main + 588
7   dyld                          	       0x199d5e0e0 start + 2360

Thread 1:
0   libsystem_pthread.dylib       	       0x19a0e1d20 start_wqthread + 0

Thread 2:
0   libsystem_pthread.dylib       	       0x19a0e1d20 start_wqthread + 0


Thread 0 crashed with ARM Thread State (64-bit):
    x0: 0x00006000038b0000   x1: 0x00006000038b0000   x2: 0x0000000000000000   x3: 0x0000000000000000
    x4: 0x00006000038b0b40   x5: 0x0000000000000ae0   x6: 0x0000000000000000   x7: 0x0000000000000000
    x8: 0x0000000000000000   x9: 0x0000000000000000  x10: 0x0000000000000000  x11: 0x0000000000000000
   x12: 0x0000000000000000  x13: 0x0000000000000000  x14: 0x0000000000000000  x15: 0x0000000000000000
   x16: 0x0000000000000000  x17: 0x0000000000000000  x18: 0x0000000000000000  x19: 0x000000010395cd50
   x20: 0x000000010395cd50  x21: 0x00006000038b0000  x22: 0x0000000000000002  x23: 0x000000011e808200
   x24: 0x0000000000000000  x25: 0x6db6db6db6db6db7  x26: 0x000000011df04890  x27: 0x000000011e808c00
   x28: 0x0000000000000000   fp: 0x000000016f844aa0   lr: 0x00000001006aa9b0
    sp: 0x000000016f844a60   pc: 0x000000010071d37c cpsr: 0x60001000
   far: 0x0000000000000000  esr: 0xf2000001 (Breakpoint) brk 1

Binary Images:
       0x100dc4000 -        0x100dc7fff libcodedirectory.1.dylib (*) <199f4d41-98bc-3ff9-9e16-3d03854facc4> /private/tmp/*/libcodedirectory.1.dylib
       0x100e58000 -        0x100eb3fff libtapi.dylib (*) <28bbd910-0af2-3662-b2ab-05a71cbec9df> /Volumes/VOLUME/*/libtapi.dylib
       0x10126c000 -        0x1014effff libcrypto.3.dylib (*) <00992c5c-6980-3a4f-9690-bffaf2bf463d> /Volumes/VOLUME/*/libcrypto.3.dylib
       0x100df4000 -        0x100e03fff libxar.1.dylib (*) <31c076eb-e55d-3337-ae62-1dcdeb6bd00a> /Volumes/VOLUME/*/libxar.1.dylib
       0x100fe8000 -        0x101093fff libc++.1.0.dylib (*) <68b05caf-2e5f-3d32-a986-3069ccf82d96> /Volumes/VOLUME/*/libc++.1.0.dylib
       0x100e14000 -        0x100e27fff libz.1.3.1.dylib (*) <e163925a-c49a-30c4-b973-ceb826394431> /Volumes/VOLUME/*/libz.1.3.1.dylib
       0x100f34000 -        0x100f63fff libc++abi.1.0.dylib (*) <84614819-b9c3-3730-bdd0-3bc3cab4b33d> /Volumes/VOLUME/*/libc++abi.1.0.dylib
       0x100f8c000 -        0x100fb3fff liblzma.5.dylib (*) <08ef1d45-21ec-3d81-be64-7ff2b1a52a59> /Volumes/VOLUME/*/liblzma.5.dylib
       0x100e34000 -        0x100e43fff libbz2.1.dylib (*) <b3ec40c0-4e92-39c4-9c6a-6257e4f4462e> /Volumes/VOLUME/*/libbz2.1.dylib
       0x101600000 -        0x10170ffff libxml2.2.dylib (*) <ef7a2e9e-9b7c-3cb7-91d7-7304c7eba479> /Volumes/VOLUME/*/libxml2.2.dylib
       0x100edc000 -        0x100ee7fff libiconv.2.dylib (*) <8f045731-0218-3975-8530-594aa9c59ac0> /Volumes/VOLUME/*/libiconv.2.dylib
       0x100dd4000 -        0x100dd7fff libcharset.1.dylib (*) <709ed437-3390-31b5-859e-f7044e79111b> /Volumes/VOLUME/*/libcharset.1.dylib
       0x1005b8000 -        0x10085bfff ld (*) <dfe4710f-d0a4-3a0f-a7fc-862bdbb20ae6> /private/tmp/*/ld
       0x199d58000 -        0x199de0a17 dyld (*) <37bbc384-0755-31c7-a808-0ed49e44dd8e> /usr/lib/dyld
               0x0 - 0xffffffffffffffff ??? (*) <00000000-0000-0000-0000-000000000000> ???
       0x19a0e0000 -        0x19a0ecfff libsystem_pthread.dylib (*) <386b0fc1-7873-3328-8e71-43269fd1b2c7> /usr/lib/system/libsystem_pthread.dylib
```

Reply:
weird -- that hit a breakpoint.

Reply:
> aarch64-darwin. This crash happened every time. I didn’t check x86_64-darwin. The crashes that happen there don’t happen reliably (i.e., you can retry the build, and it succeeds).
> 
> https://gist.github.com/reckenrode/33d69a3e0fb3b42610603b9c6c544c2c

so this looks like it is when ld64 is built in the preConfigure when it is using a debug build and asserts are enabled and thus failing on an assert -- which isn't what i was asking about. i think i hit the same thing on x64 when a `_LIBCPP_ASSERT` fires in a debug build tho the assert seems unnecessary.

```
 include/c++/v1/vector:1459: assertion __n < size() failed: vector[] index out of bounds[2]
```
but __n is zero and empty, but the calling code is going to fail if the vector is empty anyway, as `&vector[0] != &vector[0]` will fail and thus will not try to process any data in the vector.

i disabled the assertions and successfully built up to `nix-build -A git -A SDL -A SDL2` without the UB patch. I am trying to build again with the UB patch enabled to see if it fails or not.

Reply:
I dropped the UB patch and stopped building with LTO (except in a separate install check phase to check that it works). Let’s see if that helps.

> so this looks like it is when ld64 is built in the preConfigure when it is using a debug build and asserts are enabled and thus failing on an assert -- which isn't what i was asking about.

You asked about the crash, and that was the crash I was experiencing. The other crash happens when linking other things, but it’s not reproducible. Usually when it crashes, rerunning the build will successfully link.

This is one of the crash logs, but they vary: https://gist.github.com/reckenrode/a72a5b0f1376490465e5cf3ac5495776.

> but __n is zero and empty, but the calling code just checking if the vector is empty anyway, as &vector[0] != &vector[0] will fail and thus will not try to process any data in the vector.

`&vector[0]` is supposed to be UB, and clang will warn about comparing the result, but it only does so when I compare it to a `nullptr` or use it in a conditional. If I compare `&vector[0] != &vector[0]`, it doesn’t. It also doesn’t if I hide the access in a method (like Apple does in ld64).

Reply:
ok. looking at that crash dump and then disasembing the function it looks like the indirect call to atom->fixupsEnd() fails due to the atom object being corrupted and this happens before entering the loop or accessing the `&vector[end_postion]`. so something is corrupting the atom object or the atom object is garbage, if i'm reading the disassembly correctly.



```c++
void OutputFile::applyFixUps(ld::Internal& state, uint64_t mhAddress, const ld::Atom* atom, uint8_t* buffer)
{
        //fprintf(stderr, "applyFixUps() on %s\n", atom->name());
        int64_t accumulator = 0;
        const ld::Atom* toTarget = NULL;        
        const ld::Atom* fromTarget;
        int64_t delta;
        uint32_t instruction;
        uint32_t newInstruction;
#if SUPPORT_ARCH_riscv
        uint32_t imm20; 
        uint32_t imm10_1;
        uint32_t imm11;
        uint32_t imm19_12;
        uint32_t imm12;
        int32_t  signedImm12;
#endif
        bool is_bl;     
        bool is_blx;    
        bool is_b;
        bool thumbTarget = false;
        bool isRelative = false;
        std::map<uint32_t, const Fixup*> usedByHints;
#if SUPPORT_ARCH_arm64e
        Fixup::AuthData authData;
#endif  
        for (ld::Fixup::iterator fit = atom->fixupsBegin(), end=atom->fixupsEnd() /* crash here*/; fit != end; ++fit) { 

```

```asm
000000010003f390 <ld::tool::OutputFile::applyFixUps(ld::Internal&, unsigned long long, ld::Atom const*, unsigned char*)>:
10003f390: 55                           pushq   %rbp
10003f391: 48 89 e5                     movq    %rsp, %rbp
10003f394: 41 57                        pushq   %r15
10003f396: 41 56                        pushq   %r14
10003f398: 41 55                        pushq   %r13
10003f39a: 41 54                        pushq   %r12
10003f39c: 53                           pushq   %rbx
10003f39d: 48 81 ec e8 01 00 00         subq    $488, %rsp
10003f3a4: 4c 89 85 28 fe ff ff         movq    %r8, -472(%rbp)
10003f3ab: 48 89 cb                     movq    %rcx, %rbx  ; <- rcx and rbx contain pointer to atom object (4th arg (3rd + 1 for this ptr))  
10003f3ae: 48 89 95 40 fe ff ff         movq    %rdx, -448(%rbp)
10003f3b5: 48 89 b5 60 fe ff ff         movq    %rsi, -416(%rbp)
10003f3bc: 49 89 fe                     movq    %rdi, %r14
10003f3bf: 48 8b 05 c2 0c 2e 00         movq    3017922(%rip), %rax  # 100320088 <dyld_stub_binder+0x100320088>
10003f3c6: 48 8b 00                     movq    (%rax), %rax
10003f3c9: 48 89 45 d0                  movq    %rax, -48(%rbp)
10003f3cd: 48 c7 45 c8 00 00 00 00      movq    $0, -56(%rbp)
10003f3d5: 48 8d 45 b0                  leaq    -80(%rbp), %rax
10003f3d9: 0f 57 c0                     xorps   %xmm0, %xmm0
10003f3dc: 0f 11 45 b0                  movups  %xmm0, -80(%rbp)
10003f3e0: 48 89 45 a8                  movq    %rax, -88(%rbp)
10003f3e4: 48 8b 01                     movq    (%rcx), %rax  ; <-- atom into rax
10003f3e7: 48 89 cf                     movq    %rcx, %rdi
10003f3ea: ff 50 60                     callq   *96(%rax)   ; <-- call  atom->fixupsBegin()
10003f3ed: 49 89 c5                     movq    %rax, %r13  ; <-- store fit into %r13
10003f3f0: 4c 89 b5 58 fe ff ff         movq    %r14, -424(%rbp)
10003f3f7: 48 8b 03                     movq    (%rbx), %rax ; <-- atom in to rax
10003f3fa: 48 89 9d 88 fe ff ff         movq    %rbx, -376(%rbp)
10003f401: 48 89 df                     movq    %rbx, %rdi
10003f404: ff 50 68                     callq   *104(%rax)  ; <-- crash here (0x10003f390 + 116) atom->fixupsEnd()
10003f407: 48 89 85 70 fe ff ff         movq    %rax, -400(%rbp)  ;  <- save end onto the stack
10003f40e: 49 39 c5                     cmpq    %rax, %r13   ; <- check fit != end
10003f411: 0f 84 c8 1f 00 00            je      0x1000413df <ld::tool::OutputFile::applyFixUps(ld::Internal&, unsigned long long, ld::Atom const*, unsigned char*)+0x204f>
10003f417: 48 8b 8d 58 fe ff ff         movq    -424(%rbp), %rcx
10003f41e: 48 8d 81 a0 00 00 00         leaq    160(%rcx), %rax
10003f425: 48 89 85 78 fe ff ff         movq    %rax, -392(%rbp)
10003f42c: 48 8b 85 88 fe ff ff         movq    -376(%rbp), %rax
10003f433: 48 83 c0 10                  addq    $16, %rax
10003f437: 48 89 85 80 fe ff ff         movq    %rax, -384(%rbp)
10003f43e: 48 8b 85 60 fe ff ff         movq    -416(%rbp), %rax
10003f445: 48 05 28 01 00 00            addq    $296, %rax
10003f44b: 48 89 85 90 fe ff ff         movq    %rax, -368(%rbp)
10003f452: 48 8b 85 40 fe ff ff         movq    -448(%rbp), %rax
10003f459: 48 f7 d8                     negq    %rax
10003f45c: 48 89 85 10 fe ff ff         movq    %rax, -496(%rbp)
10003f463: 48 8d 81 a8 02 00 00         leaq    680(%rcx), %rax
10003f46a: 48 89 85 18 fe ff ff         movq    %rax, -488(%rbp)
10003f471: 48 8d 81 10 01 00 00         leaq    272(%rcx), %rax
10003f478: 48 89 85 08 fe ff ff         movq    %rax, -504(%rbp)
10003f47f: 31 c0                        xorl    %eax, %eax
10003f481: 48 89 85 68 fe ff ff         movq    %rax, -408(%rbp)
10003f488: 45 31 f6                     xorl    %r14d, %r14d
10003f48b: 45 31 ff                     xorl    %r15d, %r15d
10003f48e: 89 85 3c fe ff ff            movl    %eax, -452(%rbp)
10003f494: 89 85 38 fe ff ff            movl    %eax, -456(%rbp)
10003f49a: 89 85 48 fe ff ff            movl    %eax, -440(%rbp)
10003f4a0: eb 4f                        jmp     0x10003f4f1 <ld::tool::OutputFile::applyFixUps(ld::Internal&, unsigned long long, ld::Atom const*, unsigned char*)+0x161>
10003f4a2: 8b 03                        movl    (%rbx), %eax
10003f4a4: 89 c1                        movl    %eax, %ecx
10003f4a6: 81 e1 00 00 c0 bf            andl    $3217031168, %ecx
10003f4ac: 81 f9 00 00 40 b9            cmpl    $3107979264, %ecx
10003f4b2: 0f 85 61 4a 00 00            jne     0x100043f19 <ld::tool::OutputFile::applyFixUps(ld::Internal&, unsigned long long, ld::Atom const*, unsigned char*)+0x4b89>
10003f4b8: 44 89 f9                     movl    %r15d, %ecx
10003f4bb: c1 e1 0a                     shll    $10, %ecx
10003f4be: 81 e1 00 fc 3f 00            andl    $4193280, %ecx
10003f4c4: 25 ff 03 00 00               andl    $1023, %eax
10003f4c9: 01 c8                        addl    %ecx, %eax
10003f4cb: 05 00 00 00 91               addl    $2432696320, %eax
10003f4d0: 89 03                        movl    %eax, (%rbx)
10003f4d2: 66 66 66 66 66 2e 0f 1f 84 00 00 00 00 00    nopw    %cs:(%rax,%rax)
10003f4e0: 49 83 c5 10                  addq    $16, %r13   ;  <-- increment fit (lots of backward jumps to this address, so think it is the top of the loop in most cases)
10003f4e4: 4c 3b ad 70 fe ff ff         cmpq    -400(%rbp), %r13 ;  <-- check that fit != end
10003f4eb: 0f 84 ee 1e 00 00            je      0x1000413df <ld::tool::OutputFile::applyFixUps(ld::Internal&, unsigned long long, ld::Atom const*, unsigned char*)+0x204f>
...
```

Reply:
tho, as you mentioned on matrix the other day, `rip: <unavailable> ` in the registers is a little disturbing as it should be `0x102c72284`. or at least that is what i expected it to be when i dissembled the thing.

Reply:
I’m no longer experiencing the crashes after dropping LTO. I assume mixing UB and LTO didn’t work very well, causing the vtable of one of the `ld::Atom` classes to get mangled/removed.